### PR TITLE
[model_partition] Add scripts for runtime-driven, heuristic-based model partitioning  

### PR DESCRIPTION
**Pull request by dr-venkman**
[model_partition] Add scripts for runtime-driven, heuristic-based model partitioning  

Add scripts for Graph partitioning. The algorithm is a heuristic that mixes topological sorting, followed by a variant of the Fiduccia Mattheyses algorithm (see https://en.wikipedia.org/wiki/Fiduccia%E2%80%93Mattheyses_algorithm). 

Note that the data for the algorithm comes from the TFlite model and a chrome trace output from running `nnpackage_run`.

The partitions approximate a MinMax split of execution times, such that they can be pipelined.

Signed-off-by: venkat.iyer <venkat.iyer@samsung.com>

### CODE REVIEW

**Comment by juitem** [comment]
Until reading readme.md, this commit can be confused as if supporting generic model splitter.
How about copy several descriptions from readme.md  into title and commit message?
e.g) Heuristic Graph Partitioning script with chrome trace results.
...

**Comment by dr-venkman** [comment]
> Until reading readme.md, this commit can be confused as if supporting generic model splitter.
> How about copy several descriptions from readme.md into title and commit message?
> e.g) Heuristic Graph Partitioning script with chrome trace results.
> ...

Thanks for both the observation and suggestion. I have modified title and commit message accordingly.

**Comment by chunseoklee** [comment]
@dr-venkman Could you please add a list of dependent packages into README.md ?

File: tools/model_partition_tool/Graph.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
+#! /usr/bin/python
```
Comment (intom) [inline tools/model_partition_tool/Graph.py:1]: It would be nice to specify python version if it depends on version.

**Comment by dr-venkman** [comment]
> @dr-venkman Could you please add a list of dependent packages into README.md ?

Thanks for the suggestion, I will update README.md accordingly.

File: tools/model_partition_tool/Graph.py (hunk 1 above)
Comment (dr-venkman) [reply tools/model_partition_tool/Graph.py:1]: There is a strong dependency on python3 as the tflite interpreter functions to get operation details are supported only via python3 - tensorflow 2.x combination, as far as I recall.

Comment (dr-venkman) [reply tools/model_partition_tool/Graph.py:1]: I have mentioned the dependency on python3 under README.md. Please execute the scripts as `python3 <test_script.py>`

File: tools/model_partition_tool/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+This folder contains the necessary scripts to perform a a heuristic-based graph partitioning for machine learning models.
+
+The main contents of this folder are as follows:
+
+- [Python Files](#python-scripts)
+- [How to Run Partitioning Algorithm?](#how-to-partition-tflite-model)
+
+
+## Python Scripts
+The python scripts in this folder require **Python3** to retrieve TFLite model operations. Additionally, please ensure that the python `numpy` package has been installed beforehand. The scripts also import the following modules: `queue`, `json` and `argparse`, all of which should be available by default. If not, please install them either by `pip install <package>` or `sudo apt install python-<package>`.
```
Comment (chunseoklee) [inline tools/model_partition_tool/README.md:12]: @dr-venkman Do we need tensorflow package?

Comment (dr-venkman) [reply tools/model_partition_tool/README.md:12]: The script uses the tfLite Interpreter to fetch operation and tensor details to produce a adjacency matrix. The partitioning algorithm relies on the weighted adjacency matrix to produce the splits.

Comment (dr-venkman) [reply tools/model_partition_tool/README.md:12]: The script runs locally on development environment, so the requirement for TF package shouldn't be too restrictive. Please let me know your view.

Comment (chunseoklee) [reply tools/model_partition_tool/README.md:12]: @dr-venkman I do not think it's too rectitive. I mean "please add tensorflow(+verison) info into readme.md for users" :)

Comment (dr-venkman) [reply tools/model_partition_tool/README.md]: @chunseoklee , got it :), I have revised readme.md accordingly. Thanks for the comment.

**Review by chunseoklee — APPROVED** [review]
LGTM for now.

**Comment by chunseoklee** [comment]
@intom PTAL

**Review by intom — APPROVED** [review]
LGTM

**Comment by chunseoklee** [comment]
@nnfw-bot help

**Comment by chunseoklee** [comment]
@nnfw-bot format-checker

**Comment by chunseoklee** [comment]
@nnfw-bot test format-checker

File: tools/model_partition_tool/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+INFO:Topology:Session Graph:
+[[  0   1   0   0]
+ [  0   0   1   0]
+ [  0   0   0   1]
+ [  0   0   0   0]]
+INFO:Topology:Edge cut: 12
+INFO:Topology:Memory overhead (bytes): 4403136
+
+$ cat /tmp/nnpackage/inception_v3/partition_map.json
+{"partition_map": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 2, 2, 2, 3, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]}
```
Comment (intom) [inline tools/model_partition_tool/README.md:136]: It would be nice if the partition map has the information about the number of partitions.

Comment (dr-venkman) [reply tools/model_partition_tool/README.md:136]: I have added the field in the latest commit. Thank you for the comment.

**Comment by intom** [comment]
@chunseoklee PTAL

**Comment by chunseoklee** [comment]
@nnfw-bot format-checker

**Comment by chunseoklee** [comment]
@nnfw-bot test format-checker

**Comment by chunseoklee** [comment]
@dr-venkman LGTM for me. But CI checker seems not working for this PR. Could you please make a new PR with different commit sha ?

**Comment by dr-venkman** [comment]
@nnfw-bot test format-checker

**Comment by chunseoklee** [comment]
@dr-venkman We do not allow to have merge commit(576098c3f6153b74964635ac9d1892d196d6245d) on master. Could you please just rebase on the master ?

**Comment by dr-venkman** [comment]
Reopened in #7064

**Pull request closed without merging**